vpr/src/place/verify_placement.cpp

```diff
@@ -0,0 +1,253 @@
+    for (size_t imacro = 0; imacro < pl_macros.size(); imacro++) {
+        auto head_iblk = pl_macros[imacro].members[0].blk_index;
+
+        for (size_t imember = 0; imember < pl_macros[imacro].members.size(); imember++) {
+            auto member_iblk = pl_macros[imacro].members[imember].blk_index;
+
+            // Compute the supposed member's x,y,z location
+            t_pl_loc member_pos = block_locs[head_iblk].loc + pl_macros[imacro].members[imember].offset;
+
+            // Check the blk_loc_registry.block_locs data structure first
+            if (block_locs[member_iblk].loc != member_pos) {
+                VTR_LOG_ERROR(
+                    "Block %zu in pl_macro #%zu is not placed in the proper orientation.\n",
+                    size_t(member_iblk), imacro);
+                num_errors++;
+            }
+
+            // Then check the blk_loc_registry.grid data structure
+            if (grid_blocks.block_at_location(member_pos) != member_iblk) {
+                VTR_LOG_ERROR(
+                    "Block %zu in pl_macro #%zu is not placed in the proper orientation.\n",
+                    size_t(member_iblk), imacro);
+                num_errors++;
+            }
+        } // Finish going through all the members
+    } // Finish going through all the macros
+
+    return num_errors;
+}
+
+/**
+ * @brief Check that the placement is following the floorplanning constraints.
+ *
+ * This checks that all constrained blocks are placed within their floorplanning
+ * constraints.
+ *
+ *  @param blk_loc_registry     A registry containing the block locations and
+ *                              the blocks at each grid location.
+ *  @param cluster_constraints  The constraints on the placement of each
+ *                              cluster.
+ *  @param clb_nlist            The clustered netlist being verified.
+ *
+ *  @return The number of errors in the placement regarding floorplanning.
+ */
+static unsigned check_placement_floorplanning(const BlkLocRegistry& blk_loc_registry,
+                                              const vtr::vector<ClusterBlockId, PartitionRegion>& cluster_constraints,
+                                              const ClusteredNetlist& clb_nlist) {
+    const auto& block_locs = blk_loc_registry.block_locs();
+
+    unsigned num_errors = 0;
+    for (ClusterBlockId blk_id : clb_nlist.blocks()) {
+        const PartitionRegion& blk_pr = cluster_constraints[blk_id];
+        // If the cluster is not constrained, no need to check.
+        if (blk_pr.empty())
+            continue;
+        // Check if the block is placed in its constrained partition region.
+        const t_pl_loc& blk_loc = block_locs[blk_id].loc;
+        if (!blk_pr.is_loc_in_part_reg(blk_loc)) {
+            VTR_LOG_ERROR(
+                "Block %zu is not in correct floorplanning region.\n",
+                size_t(blk_id));
+            num_errors++;
+        }
```

Comment:
This is kind of my personal preference, so feel free to ignore it.

I think adding a few empty lines make the code easier to read. For example, I think of checking emptiness of `blk_pr` as a logical step. If the if statement and its comment are preceded and followed by empty lines, I can visually recongnize it as a separte step.